### PR TITLE
Support raw HTML <br> line breaks in the Markdown renderer

### LINKED ISSUE #13732: Markdown viewer: support `<br>` for explicit line breaks
### Pre-submit Checks

- [x] I have [searched Warp feature requests](https://github.com/warpdotdev/warp/issues?q=is%3Aissue+label%3Aenhancement) and there are no duplicates
- [x] I have [searched Warp docs](https://docs.warp.dev) and my feature is not there

### Describe the solution you'd like?

Warp's Markdown Viewer should convert raw `<br>` tags into an actual line break, both standalone mid-paragraph and inside GFM pipe-table cells (the most common place `<br>` shows up, since pipe-table syntax has no other way to express a multi-line cell).

**Expected behavior:** `<br>` produces a real line break wherever it appears, including inside table cells.

### Is your feature request related to a problem? Please describe.

Concretely, today in Warp's Markdown Viewer (verified on this build):

Screenshot showing the break element rendering as code

`<br>` renders as literal text in both positions — standalone mid-paragraph, and inside a GFM pipe-table cell — it is not converted to a line break in either case.

### Additional context

**Test case:**

```markdown
# `<br>` (standalone and inside a GFM pipe-table cell)

Should render: two lines separated by a real line break where the standalone `<br>` sits, and a table cell with wrapped multi-line content where the `<br>` sits inside a pipe-table cell. Currently: expect `<br>` to render as literal text in both positions — it is not converted to a line break in either the plain-paragraph case or inside GFM table cells.

Standalone, mid-paragraph:

First line of the paragraph.<br>
Second line, after a raw `<br>`.

Inside a GFM pipe-table cell:

| Feature | Notes                          |
| ------- | ------------------------------- |
| Export  | Supports CSV.<br>Also supports JSON. |
```

**Code pointers.** As far as I can tell:

- [`markdown_parser.rs`](https://github.com/warpdotdev/warp/blob/master/crates/markdown_parser/src/markdown_parser.rs)'s `InlineToken` enum has no `LineBreak` variant, so `<br>` isn't recognized anywhere in the `.md` render path — including inside GFM table cells, which route through the same `parse_inline`/`parse_cell_content` machinery as regular paragraph text.
- There's already reuse precedent for exactly this conversion: the paste-path parser ([`html_parser.rs`](https://github.com/warpdotdev/warp/blob/13902088affaf958188f99d7357cfeb7379463c4/crates/markdown_parser/src/html_parser.rs#L335)) maps `<br>` → `FormattedTextLine::LineBreak` when converting pasted HTML. Wiring the same mapping into the `.md` parser's inline tokenizer looks like the natural approach rather than inventing a new mechanism.
- CommonMark's own hard-break syntax (trailing double-space or backslash before a newline) also looks unhandled in the inline tokenizer — flagging it here since it's adjacent, though it's not the primary ask of this issue.

**Related:** this splits out one tag from #13652 (bulk raw-HTML-subset request). Sibling issues from the same split: #13721 (`` sizing), #13725 (anchor links), #13726 (raw HTML tables), #10259 (`<details>`/`<summary>` — pre-existing request), #13733 (`<kbd>`), #13734 (`<sub>`/`<sup>`), #13735 (`align` on `<p>`/`<div>`), #13736 (`<picture>`/`<source>`)

### Operating system (OS)

macOS

### How important is this feature to you?

4

### Warp Internal (ignore) - linear-label:39cc6478-1249-4ee7-950b-c428edfeecd1

_No response_

### PR DESCRIPTION
**Pull request by fbartho**
Closes #13732

The Markdown viewer rendered a raw `<br>` as literal text, both mid-paragraph and inside GFM pipe-table cells. Inside a pipe-table cell `<br>` is the only way to author a forced line break at all, so this was the more painful half of the gap.

This teaches the `.md` inline parser to recognize `<br>`, `<br/>`, and `<br />` (case-insensitive) and turn them into a real line break, in ordinary paragraphs and inside GFM table cells.

Notes:
- The break is stored as a newline fragment; the render path already turns an embedded newline into a visual break in both paragraphs and cells, so no rendering code changed. The table serialization paths translate the break back to `<br>` so an in-cell break does not corrupt row/column structure on round-trip.
- Attributes (`<br class="x">`) and malformed forms (`<br` without `>`, `< br>`, `</br>`) fall through to literal text.
- Markdown-native hard breaks (trailing double-space / backslash) are unchanged and out of scope.

Tested: new unit tests in `markdown_parser` pass, clippy and rustfmt clean, and exercised in a local dev build against the issue sample document — paragraphs, consecutive breaks, self-closing/case variants, table cells, and malformed forms all behave as described.

<details><summary>Human-focused Test Document + Screenshot</summary>

```markdown
# `<br>` line breaks (paragraphs and GFM pipe-table cells)

Each `<br>` below should render as a real line break. Malformed forms stay literal.

Mid-paragraph:

First line of the paragraph.<br>
Second line, after a raw `<br>`.

Consecutive breaks (should produce blank lines):

Before the gap.<br><br><br>After a triple `<br>`.

Self-closing and case variants (each breaks):

alpha<br/>bravo<br />charlie<BR>delta

Inside a GFM pipe-table cell (two lines, table intact):

| Feature | Notes                                |
| ------- | ------------------------------------ |
| Export  | Supports CSV.<br>Also supports JSON. |

Trailing `<br>` at the end of a cell (should add a trailing empty line inside the cell, matching GitHub):

| Feature | Notes                  |
| ------- | ---------------------- |
| Import  | Ends with a break.<br> |

Malformed / attributed forms (each should stay literal text, no break):

one</br>two &lt;br three<br class="x">four

Control (unrelated tags unaffected): a <kbd>K</kbd> stays literal.
```
Screenshot of the above test document

</details>

### CODE REVIEW

**Comment by fbartho** [comment]
### Feedback from original PR:

This PR continues #13854 (same branch history; that PR exhausted its automated-review budget). All feedback received there, and everything a deeper self-review pass surfaced, is accounted for below.

| Source | Issue | Resolution |
| --- | --- | --- |
| [Oz round 1 (inline)](https://github.com/warpdotdev/warp/pull/13854#discussion_r3599109730) | Editor's GFM table export uses a local `inline_to_markdown` that wrote raw newlines into pipe cells on copy/export, splitting rows | Same newline→`<br>` guard applied in the editor export path, with a round-trip test (ca6a4dd3) |
| [Oz round 2 (inline)](https://github.com/warpdotdev/warp/pull/13854#discussion_r3599265608), re-flagged in [round 3](https://github.com/warpdotdev/warp/pull/13854#pullrequestreview-4718117399) before the fix landed | Table cells dropped a `<br>` at a style boundary — only the final run's stripped trailing newline was preserved | Per-run newline reinsertion in `measure_table_cells`, with style-boundary tests (993de7b4) |
| [Oz round 1 (overview)](https://github.com/warpdotdev/warp/pull/13854#pullrequestreview-4717884575) | No spec context provided | #13732 is labeled `ready-to-implement` (direct code PR per CONTRIBUTING's readiness labels); round 2 recorded "no spec-drift finding" |
| self-discovered | Table-cell offset map assumed one source char per rendered char, so cursor, selection highlight, and clipboard slices mapped wrong after an in-cell break | Segment-wise walk that treats the `<br>` token as a source gap, with offset-mapping tests (0d4b4ebf) |
| self-discovered | Copy-as-HTML wrote the break as a raw newline inside `<td>`, which HTML collapses to a space | Emit a real `<br>` element, with an HTML serialization test (538c9616) |
| self-discovered | Plain-text table copy emitted a raw newline inside a tab-delimited row, so re-pasting split the row | Flavor split by design: the HTML clipboard flavor carries a `<br>` element (Warp-to-Warp paste round-trips); the plain-text flavor keeps a real newline rather than injecting HTML; trade-off documented in-test (716e5699 + 6b3ea8db) |
| self-discovered | No test coverage for `<br>` combined with style delimiters (bold/underline/links) | Regression guards added; styles carry across the break (5b346de4) |
| self-discovered | Hypothesis: a styled newline fragment might paint a stray underline/bold sliver on macOS CoreText | Manually verified in a local dev build — no artifacts at break points; no code change needed |
| self-discovered | The new (not-yet-wired) TUI table renderer counts an embedded newline into column width and would misalign a multi-line cell | ❌ out-of-scope: that renderer is currently test-only and unwired; flagging it for whoever wires it up |
| self-discovered | A cell that acquired a raw newline from a non-`<br>` source (IME, multi-line paste) would canonicalize to `<br>` on first serialization | ❌ not addressed: theoretical corner — any path that stored a raw newline in a cell already corrupted the row format before this PR; the guards narrow that hole rather than widen it |

**Comment by fbartho** [comment]
@cla-bot check

File: crates/editor/src/render/model/table_offset_map.rs [hunk 1]
```diff
@@ -306,52 +337,78 @@ impl TableCellOffsetMap {
...
                 }
-                source_idx += 1;
-            }
-
-            let visible_source_start = CharOffset::from(source_idx);
 
-            for &rendered_char in &rendered_chars {
-                if source_idx >= total_source_chars {
-                    break;
+                if segment.is_empty() {
```
[the remark this answers is not in the document]
Comment (fbartho) [reply crates/editor/src/render/model/table_offset_map.rs:361]: Okay, I’m not entirely sure what this means, so I’ll have a think on it and try to make some test cases, unless you can give me a test case that is failing @oz-for-oss?

[the remark this answers is not in the document]
Comment (fbartho) [reply crates/editor/src/render/model/table_offset_map.rs:361]: Built the failing cases!

Fixed in a7d7d2c38 by emitting a zero-length range at the source boundary for each empty segment, so every `<br>` owns its own boundary. (One nuance: the leading `<br>a` case actually already resolved correctly — kept a test for it as a guard.) Four new tests cover the edge shapes.

File: crates/markdown_parser/src/markdown_parser.rs [hunk 2]
```diff
@@ -1035,6 +1035,14 @@ fn parse_inline<'a, E: ContextError<&'a str> + ParseError<&'a str>>(
             InlineToken::UnderlineEnd => {
                 input = parse_underline(&mut state, remaining);
             }
+            InlineToken::LineBreak => {
+                // Emit the break as a newline fragment. Both the paragraph and table-cell
+                // render paths turn an embedded newline into a real visual break. The
+                // fragment may later coalesce into adjacent plain text (the newline is
+                // preserved, so rendering is unaffected); serialization paths translate any
+                // embedded newline back to `<br>`. See `HARD_LINE_BREAK`.
+                state.push_closed_node(FormattedTextFragment::plain_text(HARD_LINE_BREAK));
```
Comment (bnavetta) [inline crates/markdown_parser/src/markdown_parser.rs]: We can use `state.push_text` instead, which will avoid unnecessary coalescing of fragments later on

File: crates/markdown_parser/src/markdown_parser.rs [hunk 3]
```diff
@@ -1654,6 +1663,33 @@ fn parse_inline_token_underline_end<'a, E: ContextError<&'a str> + ParseError<&'
...
+    context(
+        "line_break",
+        value(
+            InlineToken::LineBreak,
+            tuple((
+                tag("<"),
+                tag_no_case("br"),
+                space0,
+                // Optional self-closing slash: `<br/>` or `<br />`.
+                map(take_while_m_n(0, 1, |c| c == '/'), |_| ()),
```
Comment (bnavetta) [inline crates/markdown_parser/src/markdown_parser.rs]: The `take_while_m_n` is equivalent to an `opt` combinator, I believe

File: crates/markdown_parser/src/markdown_parser_tests.rs [hunk 4]
```diff
@@ -2912,3 +2912,219 @@ fn test_parse_table_with_strikethrough() {
...
+fn test_br_in_table_cell_round_trips_as_literal_br() {
+    // A break inside a cell must serialize back to `<br>` in both the internal tab/newline
+    // format and the GFM pipe-table format, never as a raw `\n` that would split the row.
+    let source = "| A | B |\n| --- | --- |\n| one<br>two | plain |\n";
+    let result = test_parse_markdown_with_gfm_tables(source);
+    let FormattedTextLine::Table(table) = &result[0] else {
+        panic!("expected a table, got {result:?}");
+    };
+
+    let internal = table.to_internal_format();
```
Comment (bnavetta) [inline crates/markdown_parser/src/markdown_parser_tests.rs:3068]: nit: I think it'll be clearer to assert on the exact values of `internal` and `plain`, so someone reading the test can see the full expected value

File: crates/editor/src/content/markdown.rs [hunk 5]
```diff
@@ -1249,10 +1267,20 @@ fn inline_to_markdown(inline: &FormattedTextInline) -> String {
     let mut previous_styles = TextStylesWithMetadata::default();
     for fragment in inline {
         let next_styles = TextStylesWithMetadata::from(fragment.styles.clone());
+        // An authored hard break (raw HTML `<br>`) is stored as an embedded newline. Emit it as
+        // literal `<br>` so it survives the GFM pipe-table serialization without splitting the
+        // cell into a spurious extra row. Inline code cannot contain a newline from the `.md`
+        // inline parser, so it is left untouched. Mirrors the guard in
+        // `markdown_parser::inline_to_markdown`.
+        let fragment_text = if !next_styles.is_inline_code() && fragment.text.contains('\n') {
```
Comment (bnavetta) [inline crates/editor/src/content/markdown.rs]: Since (IIRC) this is a somewhat hot path, we probably don't want to allocate a new string with newlines replaced here. Instead, it should be correct to iterate over each line of `fragment.text` and call `append_formatting` / `append_content` for each. There won't be a formatting change between adjacent lines from the same fragment, but this will still make sure that we handle whitespace at the start/end of the fragment properly.

**Review by bnavetta — CHANGES_REQUESTED** [review]

File: crates/markdown_parser/src/markdown_parser.rs (hunk 2 above)
Comment (fbartho) [reply crates/markdown_parser/src/markdown_parser.rs]: Switched to push_text in 16f03d380 — it reuses the open fragment and skips the later re-coalesce. Fragment text is unchanged, so the full parser suite stays green.

File: crates/markdown_parser/src/markdown_parser.rs (hunk 3 above)
Comment (fbartho) [reply crates/markdown_parser/src/markdown_parser.rs]: Confirmed the equivalence and applied it in f256df52c as opt(char('/')).

File: crates/markdown_parser/src/markdown_parser_tests.rs (hunk 4 above)
Comment (fbartho) [reply crates/markdown_parser/src/markdown_parser_tests.rs:3068]: Done in 63a0ac6bb — both assertions now pin the exact internal and plain serializations.

File: crates/editor/src/content/markdown.rs (hunk 5 above)
Comment (fbartho) [reply crates/editor/src/content/markdown.rs]: Fixed in 31ff29c8f — it now iterates the fragment's lines and emits a literal <br> between them instead of allocating a replaced string; since styles don't change between lines, append_formatting only does work at the fragment boundary while per-line whitespace handling is preserved. Added an edge-case round-trip test (leading/trailing/consecutive breaks and a break before leading whitespace) that passed against the old path before the change, so output is byte-identical.

**Comment by fbartho** [comment]
Thanks @bnavetta! I appreciate the feedback.

I applied and tested the changes locally, so I'm going to ask for an /oz-review (since I believe that's the correct process?).

I apologize if this has been a noisy PR, I think I'm starting to get a handle on working on/with Warp and Oz! :P

**Comment by fbartho** [comment]
(Just commenting as this is waiting for review).

**Comment by fbartho** [comment]
(Just commenting as this is waiting for review).

**Review by bnavetta — APPROVED** [review]

**Comment by bnavetta** [comment]
Sorry for the delay! This looks good

**Comment by fbartho** [comment]
@oss-maintainers — This was approved but not merged. Would you like anything else?

**Comment by bnavetta** [comment]
Hmm, I’m not sure why this didn’t auto-merge. Would you be able to resolve the conflicts and I’ll re-approve?

**Comment by fbartho** [comment]
Working on that now!

**Comment by fbartho** [comment]
@bnavetta Conflicts resolved! I'll watch to see if CI is grumpy, but I don't expect any issues.

**Comment by fbartho** [comment]
@bnavetta CI is green (I'm worried about this catching new conflicts soon), what else can I do to help here?

**Comment by bnavetta** [comment]
Hey sorry! I resolved the latest conflict, will merge once CI finishes

**Review by bnavetta — APPROVED** [review]

File: crates/editor/src/content/edit_tests.rs [hunk 6]
```diff
@@ -927,6 +927,156 @@ fn test_layout_table_block_caches_cell_text_frames() {
...
+
+/// A trailing `<br>` at the end of a table cell must preserve the trailing newline in the cell's
+/// measured text so it renders an extra empty line, matching GitHub's rendering of a trailing hard
+/// break inside a GFM pipe-table cell.
+#[test]
+fn test_measure_table_cells_trailing_br_keeps_newline() {
+    App::test((), |app| async move {
+        app.read(|ctx| {
+            let layout_cache = LayoutCache::new();
+            let text_layout = TextLayout::new(
```
[the remark this answers is not in the document]
Comment (fbartho) [reply crates/editor/src/content/edit_tests.rs:915]: Addressed in 5f19e526e.

File: crates/markdown_parser/src/markdown_parser_tests.rs [hunk 7]
```diff
@@ -2952,13 +3102,52 @@ fn test_parse_html_comment_multi_line_block_is_stripped() {
 }
 
 #[test]
-fn test_parse_unterminated_html_comment_stays_literal() {
```
[the remark this answers is not in the document]
Comment (fbartho) [reply crates/markdown_parser/src/markdown_parser_tests.rs:2955]: Addressed in 97a124701.

File: crates/editor/src/render/model/table_offset_map.rs [hunk 8]
```diff
@@ -306,52 +337,93 @@ impl TableCellOffsetMap {
...
```
[the remark this answers is not in the document]
Comment (fbartho) [reply crates/editor/src/render/model/table_offset_map.rs:361]: Addressed in c53adca40.

**Comment by fbartho** [comment]
@bnavetta Since your approval on September 8, three things changed on this branch:

- Merged master again and updated the five `measure_table_cells` tests to the current three-argument `TextLayout::new` (5f19e526e). That was the compile failure in the last run.
- Restored `test_parse_unterminated_html_comment_stays_literal`, which an earlier master merge had dropped (97a124701).
- Fixed the table offset map for a `<br>` that directly follows a closing style marker, e.g. `**a**<br>b`, so selections on that break no longer land inside the markers, with tests for the sibling shapes (c53adca40).

Oz's three review triggers are used up on this pull request, so its "changes requested" review stays until a human dismisses it, and the GitHub Actions runs on the new commits are waiting on approval. Could you approve the runs and, if they come back green, dismiss the bot review or merge?
